utils.py

```diff
@@ -126,11 +126,11 @@ def get_service_ports(self, app):
                 if ports:
                     return list(ports)
 
-        ports = app.get('ports', [])
```

Comment:
Your example doesn't look like it recursively removes None.

I'm about at the limit of my python skills here. If you want additional changes made here, I kindly request that you commandeer this PR @nlsun